testbed/tests/e2e_test.go

```diff
@@ -84,43 +84,45 @@ func TestBallastMemory(t *testing.T) {
-			receiver,
-			cp,
-			&testbed.PerfTestValidator{},
-			performanceResultsSummary,
-			testbed.WithSkipResults(),
-			testbed.WithResourceLimits(testbed.ResourceSpec{ExpectedMaxRAM: test.maxRSS}),
-		)
-		tc.StartAgent()
-
-		var rss, vms uint32
-		// It is possible that the process is not ready or the ballast code path
-		// is not hit immediately so we give the process up to a couple of seconds
-		// to fire up and setup ballast. 2 seconds is a long time for this case but
-		// it is short enough to not be annoying if the test fails repeatedly
-		tc.WaitForN(func() bool {
-			rss, vms, _ = tc.AgentMemoryInfo()
-			return vms > test.ballastSize
-		}, time.Second*2, fmt.Sprintf("VMS must be greater than %d", test.ballastSize))
-
-		// https://github.com/open-telemetry/opentelemetry-collector/issues/3233
-		// given that the maxRSS isn't an absolute maximum and that the actual maximum might be a bit off,
-		// we give some room here instead of failing when the memory usage isn't that much higher than the max
-		lenientMax := 1.1 * float32(test.maxRSS)
-		assert.LessOrEqual(t, float32(rss), lenientMax)
-		cleanup()
-		tc.Stop()
+		t.Run(fmt.Sprintf("ballast-size-%d", test.ballastSize), func(t *testing.T) {
+			sender := testbed.NewOTLPTraceDataSender(testbed.DefaultHost, testbed.GetAvailablePort(t))
+			receiver := testbed.NewOTLPDataReceiver(testbed.GetAvailablePort(t))
+			ballastCfg := createConfigYaml(
+				t, sender, receiver, resultDir, nil,
+				map[string]string{"memory_ballast": fmt.Sprintf(ballastConfig, test.ballastSize)})
+			cp := testbed.NewChildProcessCollector()
+			cleanup, err := cp.PrepareConfig(ballastCfg)
+			require.NoError(t, err)
+			tc := testbed.NewTestCase(
+				t,
+				dataProvider,
+				sender,
+				receiver,
+				cp,
+				&testbed.PerfTestValidator{},
+				performanceResultsSummary,
+				testbed.WithSkipResults(),
+				testbed.WithResourceLimits(testbed.ResourceSpec{ExpectedMaxRAM: test.maxRSS}),
+			)
+			tc.StartAgent()
+
+			var rss, vms uint32
+			// It is possible that the process is not ready or the ballast code path
+			// is not hit immediately so we give the process up to a couple of seconds
+			// to fire up and setup ballast. 2 seconds is a long time for this case but
+			// it is short enough to not be annoying if the test fails repeatedly
+			tc.WaitForN(func() bool {
+				rss, vms, _ = tc.AgentMemoryInfo()
+				return vms > test.ballastSize
+			}, time.Second*2, fmt.Sprintf("VMS must be greater than %d", test.ballastSize))
+
+			// https://github.com/open-telemetry/opentelemetry-collector/issues/3233
+			// given that the maxRSS isn't an absolute maximum and that the actual maximum might be a bit off,
+			// we give some room here instead of failing when the memory usage isn't that much higher than the max
+			lenientMax := 1.1 * float32(test.maxRSS)
+			assert.LessOrEqual(t, float32(rss), lenientMax, fmt.Sprintf("The RSS memory usage (%d) is >10%% higher than the limit (%d).", rss, test.maxRSS))
```

Comment:
In theory, the parameters starting from the fourth position would be used in Sprintf, but it wasn't working:


```
    assertion_compare.go:342: 
        	Error Trace:	e2e_test.go:123
        	Error:      	"100" is not less than or equal to "99"
        	Test:       	TestBallastMemory/ballast-size-500
        	Messages:   	[The RSS memory usage (%d) is >10%% higher than the limit (%d). 83 90]
```

So, I'm doing the Sprintf manually here, which results in this:

```
=== RUN   TestBallastMemory/ballast-size-500
2021/12/10 11:21:28 Starting Agent (/home/jpkroehling/Projects/src/github.com/open-telemetry/opentelemetry-collector-contrib/bin/otelcontribcol_linux_amd64)
2021/12/10 11:21:28 Writing Agent log to /home/jpkroehling/Projects/src/github.com/open-telemetry/opentelemetry-collector-contrib/testbed/tests/results/TestBallastMemory/ballast-size-500/agent.log
2021/12/10 11:21:28 Agent running, pid=1049488
    assertion_compare.go:342: 
        	Error Trace:	e2e_test.go:123
        	Error:      	"100" is not less than or equal to "99"
        	Test:       	TestBallastMemory/ballast-size-500
        	Messages:   	[The RSS memory usage (100) is >10% higher than the limit (90).]
```

Reply:
Use `assert.LessOrEqual[F]`

Reply:
I'm sure I'm doing something stupid, as I'm still not getting it right:

```golang
assert.LessOrEqualf(t, float32(100), lenientMax, "The RSS memory usage (%d) is >10%% higher than the limit (%d).", rss, test.maxRSS)
```

Yields this:

```
2021/12/10 20:17:17 Agent running, pid=1315703
    assertion_compare.go:342: 
                Error Trace:    e2e_test.go:123
                Error:          "100" is not less than or equal to "99"
                Test:           TestBallastMemory/ballast-size-500
                Messages:       [The RSS memory usage (%d) is >10%% higher than the limit (%d). 83 90]
```

Reply:
I've just tried the example from their documentation, and it also doesn't work here:

```
assert.LessOrEqualf(t, 2, 1, "error message %s", "formatted")
```

Yields:
```
                Messages:       [error message %s formatted]
```